feat(security): Add documentation about add-on service section (#405)
edgexfoundry/edgex-docs: docs_src/security/Ch-Configuring-Add-On-Services.md

The settings and service environment variables are pre-wired and ready to run in secure mode without any update
or modification to the Docker-compose files.  However, there are some pre-built add-on services like `device-camera`,
`device-modbus` and some of application services (eg. `app-http-export`) are not pre-wired by default
and thus needs some deployment efforts to make them run in secure mode.

EdgeX provides a way for a user to add and configure those add-on services into EdgeX Docker
software stack running in secure mode.  This can be done vai Docker-compose files with a few addition of
environment variables and some modification of micro-service's Dockerfile.

One of the major security features in EdgeX Ireland release is to utilize the service `security-bootstrapper`
to ensure the right starting sequence so that the secrets are pre-seeded and in a more predictable starting order and
inter-dependencies between micro-services.

Currently EdgeX uses Vault as the default implementation for secret store and Consul as the configuration
and/or registry server if user chooses to do so.  There are some default services configured to generate

> secret tokens by default like those EdgeX core services and device-virtual service. There will be time of

Reviewer: This last sentence is awkward.

Author: i'll re-write the last sentence. Thanks

Author: re-written

A good example of this will be like `app-serive-rules`:

```yaml
...
  app-service-rules:
    entrypoint: ["/edgex-init/ready_to_run_wait_install.sh"]
    command: "/app-service-configurable ${DEFAULT_EDGEX_RUN_CMD_PARMS}"
    volumes:
      - edgex-init:/edgex-init:ro,z
      - /tmp/edgex/secrets/app-rules-engine:/tmp/edgex/secrets/app-rules-engine:ro,z
    depends_on:
      - security-bootstrapper
...

```

## Configure the secret store to use

Make sure the TOML configuration file of add-on service like `device-camera` contains
the proper `[SecretStore]` section.

Example:

```toml
[SecretStore]
Type = 'vault'
Host = 'localhost'
Port = 8200
Path = 'my-service/'
Protocol = 'http'
RootCaCertPath = ''
ServerName = ''
TokenFile = '/tmp/edgex/secrets/my-service/secrets-token.json'
  [SecretStore.Authentication]
  AuthType = 'X-Vault-Token'
```

and then in the EdgeX service `secretstore-setup` environment section of `Docker-compose` file to add
the service key or hostname to the environment variable `ADD_SECRETSTORE_TOKENS`:

```yaml
...
  secretstore-setup:
    container_name: edgex-secretstore-setup
    depends_on:
    - security-bootstrapper
    - vault
    environment:

>       ADD_SECRETSTORE_TOKENS: 'device-camera'

Reviewer: Should this be `my-service` to match the above instead of `device-camera`?

Author: good catch, i'll change `my-service` to `device-camera` to avoid conflicts and confusing.

Author: changed.

...

```

With that, `secretstore-setup` then can generate secretstore token from `Vault` and store it in
the `TokenFile` path specified in the TOML configuration file like the above example.

## Configure the secure Redis message bus to use

This can be done in the EdgeX service `secretstore-setup` environment section of

> `Docker-compose` file to add the `redisdb` and the service key using it.

Reviewer: ```suggestion
`docker-compose` file to add the `redisdb` and the service key using it.
```

Author: fixed

So an example will be like `redisdb[device-virtual]` in which `redisdb` is the database resource name
and the `device-virtual` is the service key or hostname and add that value into the environment variable

> `ADD_KNOWN_SECRETS`:

Reviewer: Still not clear enough what is going on.

Author: added more description for it

```yaml
...
  secretstore-setup:
    container_name: edgex-secretstore-setup
    depends_on:
    - security-bootstrapper
    - vault
    environment:

>       ADD_SECRETSTORE_TOKENS: 'device-camera, my-service'

Reviewer: Whitespace is ok?

Author: yes

      ADD_KNOWN_SECRETS: redisdb[app-rules-engine],redisdb[device-rest],redisdb[device-virtual]
...

```

> ## (Optional) Configure the ACL role of configuration/registry to use if the service depends on it

Reviewer: Since almost everyone uses Consul, shouldn't this be "Suggested" or "Recommended" instead?

Author: No. What I meant is the configuring of ACL role registry is optional if the add-on service doesn't depend on Consul although `Consul` can still be "Suggested" or "Recommended".

This is a new step coming from `securing Consul` security features as part of EdgeX Ireland release.

If the add-on service uses `Consul` as the configuration and/or registry service, then we also need to
configure the environment variable `ADD_REGISTRY_ACL_ROLES` to tell `security-bootstrapper` to generate
an ACL role for `Consul` to associate with its token.

An example of configuring acl roles of the registry `Consul` for the add-on services
`device-modbus` and `app-http-export` is like the following:

```yaml
...
  consul:
    container_name: edgex-core-consul
    depends_on:
    - security-bootstrapper
    - vault
    entrypoint:
    - /edgex-init/consul_wait_install.sh
    environment:
      ADD_REGISTRY_ACL_ROLES: app-http-export,device-modbus
...

```

> 

Reviewer: It might be useful to describe what the user might observe if they were supposed to do this but didn't.

Author: good point, i'll add that

Author: added

## Configure the API gateway access route for add-on service

If it is desirable to let user or other application services outside EdgeX's Docker network access
the endpoint of the add-on service, then we can configure and add it via the environment variable of
service `proxy-setup`, `ADD_PROXY_ROUTE`.  `proxy-setup` adds those services listed in that environment
variable into the API gateway (aka Kong) route so that the endpoint can be accessible using Kong's proxy
endpoint.

One example of adding API gateway access route for service `device-camera` is give as follows:

```yaml
...
edgex-proxy:
      ...
    environment:
      ...
      ADD_PROXY_ROUTE: "device-camera.http://device-camera:59985"
      ...
...
```

> where port number `59985` is the internal port used by service `device-camera`.

Reviewer: There is a little bit more going on here that needs to be explained, like the first part is a URL path, and the second part is a hostname, and maybe that we recommend they normally agree.

Author: ok, that details you would like to see is actually in [API Gateway documentation page](Ch-APIGateway.md) page, so I kind of briefly just mentioning here to avoid the duplication.

Author: added some brief explanation about the format of value

With that setup, we can then access the endpoints of `device-camera` from Kong's host like
`https://kong:8443/device-camera/{device-name}/name` assuming the caller can resolve `kong` from DNS server.

For more details on how the introduction to the API gateway and how it works,

> please see [Ch-APIGateway documentation page](Ch-APIGateway.md).

Reviewer: ```suggestion
please see [API Gateway documentation page](Ch-APIGateway.md).
```

Author: fixed

# Configuring Add-on Service

In the current EdgeX security serivces, we set up and configure all security related properties and environments
for the existing default serivces like `core-data`, `core-metadata`, `device-virtual`, and so on.

The settings and service environment variables are pre-wired and ready to run in secure mode without any update
or modification to the Docker-compose files.  However, there are some pre-built add-on services like `device-camera`,
`device-modbus` and some of application services (eg. `app-http-export`) are not pre-wired by default

> and thus needs some deployment efforts to make them run in secure mode.

Author: fixed

EdgeX provides a way for a user to add and configure those add-on services into EdgeX Docker
software stack running in secure mode.  This can be done vai Docker-compose files with a few addition of

> environment variables and some modification of micro-service's Dockerfile.

Author: good point. will add that

> environment variables and some modification of micro-service's Dockerfile.

Author: added

One of the major security features in EdgeX Ireland release is to utilize the service `security-bootstrapper`
to ensure the right starting sequence so that the secrets are pre-seeded and in a more predictable starting order and
inter-dependencies between micro-services.

Currently EdgeX uses `Vault` as the default implementation for secret store and Consul as the configuration
and/or registry server if user chooses to do so.  There are some default services configured to generate

> secret tokens by default like those EdgeX core services and device-virtual service.

Author: fixed

For running those add-on services (e.g. `device-camera`, `app-http-export`) in secure mode,
their secret store tokens are not generated by default from the existing docker-compose deployment

> but they can be generated through some configuring steps as shown below.

Author: fixed

In the following scenario, we assume the EdgeX services are running in Docker environments,
and thus the examples are given in terms of Docker-compose ways.  It should not be much or bigger
difference for `snap` running environment to apply the same steps or concepts if found to do so.

If a user wants to configure and set up an add-on service, e.g. `device-camera`,
he can do the steps of guidelines as follows to achieve that:

## Make add-on services security-bootstrapper compatible

> To use the Docker entrypoint scripts of gating mechanism from `security-bootstrapper`,

Author: fixed

the Dockerfile of `device-camera` should inherit shell scripting capability like `alpine`-based
as the base Docker image and should install `dumb-init`(see details in
[Why you need an init system](https://github.com/Yelp/dumb-init#why-you-need-an-init-system))
via `apk add --update` command.

Dockerfile example using alpine-base image and add `dumb-init`:

```dockerfile
......
FROM alpine:3.12

# dumb-init needed for injected secure bootstrapping entrypoint script when run in secure mode.
RUN apk add --update --no-cache dumb-init
......

```

and then in the service itself should add `/edgex-init/ready_to_run_wait_install.sh` as the entrypoint script
for the service in gating fashion and related Docker volumes from `edgex-init` and secretstore token which
will be outlined in the next section.

A good example of this will be like `app-serive-rules`:

```yaml
...
  app-service-rules:
    entrypoint: ["/edgex-init/ready_to_run_wait_install.sh"]

>     command: "/app-service-configurable ${DEFAULT_EDGEX_RUN_CMD_PARMS}"

Author: ok

>     command: "/app-service-configurable ${DEFAULT_EDGEX_RUN_CMD_PARMS}"

Author: added the why

    volumes:
      - edgex-init:/edgex-init:ro,z
      - /tmp/edgex/secrets/app-rules-engine:/tmp/edgex/secrets/app-rules-engine:ro,z
    depends_on:
      - security-bootstrapper
...

```

## Configure the secret store to use

Make sure the TOML configuration file of add-on service like `device-camera` contains
the proper `[SecretStore]` section.

Example:

```toml
[SecretStore]
Type = 'vault'
Host = 'localhost'
Port = 8200
Path = 'device-camera/'
Protocol = 'http'
RootCaCertPath = ''
ServerName = ''

> TokenFile = '/tmp/edgex/secrets/device-camera/secrets-token.json'

Author: added

  [SecretStore.Authentication]
  AuthType = 'X-Vault-Token'
```

and then in the EdgeX service `secretstore-setup` environment section of `docker-compose` file to add
the service key or hostname to the environment variable `ADD_SECRETSTORE_TOKENS`:

```yaml
...
  secretstore-setup:
    container_name: edgex-secretstore-setup
    depends_on:
    - security-bootstrapper
    - vault
    environment:

>       ADD_SECRETSTORE_TOKENS: 'device-camera'

Author: added

...

```

With that, `secretstore-setup` then can generate secretstore token from `Vault` and store it in
the `TokenFile` path specified in the TOML configuration file like the above example.

> ## Configure the secure Redis message bus to use

Author: ok since you suggest changing the section title, now this makes sense to the section I outline below.  However, this creates some confusion of securing Redis message bus for user.  Why we want to refer all of that to this and call it securing Redis message bus. If I were the user, I got lost about the relationship to convey between the "securing redis message bus" and this whole page.

Thoughts?

> ## Configure the secure Redis message bus to use

Reviewer 2: We are not `securing` the Redis MessageBus. 
We are explaining the steps an add-on service needs to do to use the already `secured` MessageBus. This is just one of those steps. 
If you want to add numbers to the steps, we could reference them in the Secure MessageBus document, i.e. Steps 1 & 2 from this document are required for the add-on service to connect to the secure MessageBus.

> ## Configure the secure Redis message bus to use

Author: ok, thanks for the explanation.  the number steps are not working well in this documentation, I 'll leave as it is in terms of sections.   
So in the "Secure Message Bus" documentation we refer to these documentation and all the sections are steps here to configure to use "secure message bus" so to speak. 
I see that in the other documentation but not here hence kind of getting lost originally.  
maybe adding some sentences at the very beginning (the introduction section of this doc) to mention that probably would make that connection better.

> ## Configure the secure Redis message bus to use

Reviewer 2: This documentation is the big picture of adding an ad-on service. The steps needed for connecting to the secure MessageBus are just creating the secret store and adding the `redisdb` known secret (which is not needed for those not connecting to the bus or needing the DB),

> ## Configure the secure Redis message bus to use

Author: sounds good, we can add what you said in this section which is actually an important concept convey to user

> ## Configure the secure Redis message bus to use

Author: changed

The default internal database of EdgeX micro-services is `Redis` and to use Redis message bus,
we can configure it the EdgeX service `secretstore-setup` environment section of
`docker-compose` file to add the `redisdb` and the service key using it.

So given an example for service `device-virtual` to use the `Redis` message bus in secure mode,
we need to tell the `secretstore-setup` to have the service "known" for the secret path of
the Redis credentials (path for hashed password basically). This can be done through
the configuration of adding `redisdb[device-virtual]` into the environment variable
`ADD_KNOWN_SECRETS` in the service `secretstore-setup`, in which `redisdb` is the
database resource name and the `device-virtual` is the service key or hostname.

```yaml
...
  secretstore-setup:
    container_name: edgex-secretstore-setup
    depends_on:
    - security-bootstrapper
    - vault
    environment:
      ADD_SECRETSTORE_TOKENS: 'device-camera, my-service'

>       ADD_KNOWN_SECRETS: redisdb[app-rules-engine],redisdb[device-rest],redisdb[device-virtual]

Author: sure, i can mention that too in additional to this

>       ADD_KNOWN_SECRETS: redisdb[app-rules-engine],redisdb[device-rest],redisdb[device-virtual]

Author: added

...

```

In the above `docker-compose` section of `secretstore-setup`, we specify the known secrets of

> `Redis` database credentials to add to services `app-rules-engine`, `device-rest`,

Reviewer: Maybe something along the lines of "copies the redisdb secret into device-rest's secret store."

Author: modified.

If the add-on service uses `Consul` as the configuration and/or registry service, then we also need to
configure the environment variable `ADD_REGISTRY_ACL_ROLES` to tell `security-bootstrapper` to generate
an ACL role for `Consul` to associate with its token.

An example of configuring ACL roles of the registry `Consul` for the add-on services
`device-modbus` and `app-http-export` is shown as follows:

```yaml
...
  consul:
    container_name: edgex-core-consul
    depends_on:
    - security-bootstrapper
    - vault
    entrypoint:
    - /edgex-init/consul_wait_install.sh
    environment:
      ADD_REGISTRY_ACL_ROLES: app-http-export,device-modbus
...

```

The configuration of Edgex service `consul`'s environment variable `ADD_REGISTRY_ACL_ROLES` tells
the `security-bootstrapper` to set up `Consul` ACL role so that the ACL token is generated,
hence the permission is granted for that service with the access to `Consul` in secure mode.

The API endpoints to Consul for the service would have been forbidden (HTTP status code = 403)
if this step was not been done due to missing configuration of ACL role and token.

> ## Configure the API gateway access route for add-on service

Author: added

If it is desirable to let user or other application services outside EdgeX's Docker network access
the endpoint of the add-on service, then we can configure and add it via the environment variable of
service `proxy-setup`, `ADD_PROXY_ROUTE`.  `proxy-setup` adds those services listed in that environment
variable into the API gateway (aka Kong) route so that the endpoint can be accessible using Kong's proxy
endpoint.

One example of adding API gateway access route for service `device-camera` is give as follows:

```yaml
...
edgex-proxy:
      ...
    environment:
      ...

>       ADD_PROXY_ROUTE: "device-camera.http://device-camera:59985"

Reviewer 2: ```suggestion
      ADD_PROXY_ROUTE: "device-camera.http://edgex-device-camera:59985"
```

>       ADD_PROXY_ROUTE: "device-camera.http://device-camera:59985"

Author: again this is a duplication of the other documentation already. but yeah i can add note to that

>       ADD_PROXY_ROUTE: "device-camera.http://device-camera:59985"

Reviewer 2: I think an example with two services demonstrates it which out have to explain it again.

>       ADD_PROXY_ROUTE: "device-camera.http://device-camera:59985"

Author: yea, examples are always illustrative i like it

>       ADD_PROXY_ROUTE: "device-camera.http://device-camera:59985"

Author: updated the example and sentences and notes

      ...
...
```

where the first part of configured value `device-camera` is the service key

> and the URL format is the service's hostname with its internal port number `59985`

Reviewer 2: ```suggestion
and the URL format is the service's hostname with its docker network port number `59985`
```

for `device-camera`.

With that setup, we can then access the endpoints of `device-camera` from Kong's host like

> `https://kong:8443/device-camera/{device-name}/name` assuming the caller can resolve `kong` from DNS server.

Reviewer 2: ```suggestion
`https://<HostName>:8443/device-camera/{device-name}/name` assuming the caller can resolve `<HostName>` from DNS server.
```

Author: fixed

# Configuring Add-on Service

In the current EdgeX security serivces, we set up and configure all security related properties and environments
for the existing default serivces like `core-data`, `core-metadata`, `device-virtual`, and so on.

The settings and service environment variables are pre-wired and ready to run in secure mode without any update
or modification to the Docker-compose files.  However, there are some pre-built add-on services like `device-camera`,

> `device-modbus` and some of application services (eg. `app-http-export`) are not pre-wired by default

Author: not totally agree with the suggestion; app-http-export is not the only one not pre-wired.  @lenny-intel 
if we put this way it may be kind of misleading example

> `device-modbus` and some of application services (eg. `app-http-export`) are not pre-wired by default

Reviewer 2: `like` infers these are not the only ones just like  `device-camera`, `device-modbus`  are not the only device service.

and thus need some configuration efforts to make them run in secure mode.

> 

Author: added

```toml
[SecretStore]
Type = 'vault'
Host = 'localhost'
Port = 8200
Path = 'device-camera/'
Protocol = 'http'
RootCaCertPath = ''
ServerName = ''
TokenFile = '/tmp/edgex/secrets/device-camera/secrets-token.json'
  [SecretStore.Authentication]
  AuthType = 'X-Vault-Token'
```

Note that the service key `device-camera` must be used for the `Path` and in the `TokenFile` path
to keep it consistent and easier to maintain.
And then add the add-on service's service key to EdgeX service `secretstore-setup`'s
`ADD_SECRETSTORE_TOKENS` environment variable in the `environment` section of `docker-compose`
as the example shown below:

```yaml
...
  secretstore-setup:
    container_name: edgex-secretstore-setup
    depends_on:
    - security-bootstrapper
    - vault
    environment:
      ADD_SECRETSTORE_TOKENS: 'device-camera'
...

```

With that, `secretstore-setup` then will generate `Secret Store` token from `Vault` and store it in
the `TokenFile` path specified in the TOML configuration file like the above example.

Also note that the value of `ADD_SECRETSTORE_TOKENS` can take more than one service in a form of
comma separated list like "`device-camera`, `device-modbus`" if needed.

> ## Configure known secrets add-on services need

Reviewer 2: ```suggestion
## (Optional) Configure known secrets for add-on services
```
```

The `ADD_KNOWN_SECRETS` environment variable on `secretstore-setup` allows for known secrets
to be added to an add-on service's `Secret Store`.

For the Ireland release, the only `known` secret is the `Redis DB credentials` identified by
the name `redisdb`. Any add-on service needing access to the `Redis DB` such as
App Service HTTP Export with Store and Forward enabled will need the `Redis DB credentials`
put in its `Secret Store`. Also, since the `Redis DB` service is now used for the MessageBus implementation,
all services that connect to the MessageBus also need the `Redis DB credentials`

> Note that the steps needed for connecting add-on services to the `secure MessageBus` are just
> creating the `Secret Stores` and then adding the `redisdb`'s known secrets. It is not necessary to do
> if the add-on service is not connecting to the bus or the Redis database.

Reviewer 2: ```suggestion
Note that the steps needed for connecting add-on services to the `Secure MessageBus` are:

1. Utilizing the security bootstrapper to ensure proper start up sequence.
2. Creating the `Secret Store` for the add-on service
3. Adding the `redisdb`'s known secret to the add-on service's `Secret Store` 

if the add-on service is not connecting to the bus or the Redis database, adding the known secret can be skipped.
```
